File: python/stack.py
class Stack:
    """
    Implements a stack data structure with a push() function, pop() function,
    and support for len(). Internally, a list is used to store the stack items.
    """

    """
    Creates a list to hold the items.
    """
    def __init__(self, items: list = []):
        self.items = items
    
    """
    Adds an item to the stack and returns the new list length.
    """
    def push(self, *args):
        self.items.append(args)

        return len(self.items)
    
    """
    Removes an item from the stack and returns it.
    """
    def pop(self):
        return self.items.pop()
    
    """
    Gives the number of items in the stack.
    """
    def __len__(self):
        return len(self.items)

File: javascript/stack.js
import {SinglyLinkedNode} from "./singly_linked_list.js";

class Stack
{
    /**
     * Creates a stack that holds one item with [data].
     *
     * @param {*} data
     */
    constructor(data)
    {
        this.head = new SinglyLinkedNode(data);

        this.length = 1;
    }

    /**
     * Pushes a new item with [data] on the stack.
     * 
     * @param {*} data 
     */
    push(data)
    {
        let new_node = new SinglyLinkedNode(data, this.head);

        this.head = new_node;

        this.length++;
    }

    /**
     * Pops an item from the stack.
     * 
     * @throws {RangeError} For when there are no items.
     */
    pop()
    {
        if (this.length === 0)
        {
            throw new RangeError(`Can only pop items if length is more than 0.`);
        }

        this.head = this.head.next;

        this.length--;
    }
}

